kms/tpmkms/tpmkms.go

```diff
@@ -0,0 +1,523 @@
+			// behavior of the built-in attestorClient, but at least it does provide
+			// a basic extension point for other ways of performing attestation that
+			// might be useful for testing or attestation flows against other systems.
+			// For it to be truly useful, the logic for determining the AK identity
+			// would have to be updated too, though.
+			ac = req.AttestationClient
+		} else {
+			ac, err = k.newAttestorClient(ek, ak)
+			if err != nil {
+				return nil, fmt.Errorf("failed creating attestor client: %w", err)
+			}
+		}
+		// perform the attestation flow with a (remote) attestation CA
+		if akChain, err = ac.Attest(ctx); err != nil {
+			return nil, fmt.Errorf("failed performing AK attestation: %w", err)
+		}
+		// store the result with the AK, so that it can be reused for future
+		// attestations.
+		if err := ak.SetCertificateChain(ctx, akChain); err != nil {
+			return nil, fmt.Errorf("failed storing AK certificate chain: %w", err)
+		}
+	}
+
+	// when a new certificate was issued for the AK, it is possible the
+	// certificate that was issued doesn't include the expected and/or required
+	// identity, so this is checked before continuing.
+	if !hasValidIdentity(ak, ekKeyURL) {
+		return nil, fmt.Errorf("AK certificate (chain) not valid for EK %q", ekKeyURL)
+	}
+
+	// prepare the response to return
+	akCert := akChain[0]
+	permanentIdentifier := ekKeyURL.String() // TODO(hs): should always match the valid value of the AK identity (for now)
+	resp := &apiv1.CreateAttestationResponse{
+		Certificate:         akCert,
+		CertificateChain:    akChain, // TODO(hs): should this include the leaf or not?
+		PublicKey:           akCert.PublicKey,
+		PermanentIdentifier: permanentIdentifier,
+	}
+
+	return resp, nil
+}
+
+// Close releases the connection to the TPM.
+func (k *TPMKMS) Close() (err error) {
+	return
+}
+
+// getPreferredEK returns the first RSA TPM EK found. If no RSA
+// EK exists, it returns the first ECDSA EK found.
+func getPreferredEK(eks []*tpm.EK) (ek *tpm.EK) {
+	var fallback *tpm.EK
+	for _, ek = range eks {
+		if _, isRSA := ek.Public().(*rsa.PublicKey); isRSA {
+			return
+		}
+		if fallback == nil {
+			fallback = ek
+		}
+	}
+	return fallback
+}
+
+// hasValidIdentity indicates if the AK has an associated certificate
+// that includes a valid identity. Currently we only consider certificates
+// that encode the TPM EK public key ID as one of its URI SANs, which is
+// the default behavior of the Smallstep Attestation CA.
+func hasValidIdentity(ak *tpm.AK, ekURL *url.URL) bool {
+	chain := ak.CertificateChain()
+	if len(chain) == 0 {
+		return false
+	}
+	akCert := chain[0]
+
+	// TODO(hs): before continuing, add check if the cert is still valid?
+
+	// the Smallstep Attestation CA will issue AK certifiates that
+	// contain the EK public key ID encoded as an URN by default.
+	for _, u := range akCert.URIs {
+		if strings.EqualFold(ekURL.String(), u.String()) {
```

Comment:
base64 is case-sensitive. Go's url.Parse always stores the schema in lower case, the only problem might be `sha256`, and I think it's ok to just use `==`

Reply:
Fixed in 1686d1023a78d323f86c7ed848a5649617611e56. 

Currently no additional logic for `ek:sha256:`, but I could add that.